File: src/data/codes/kmeans.m
clc;
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'kmeans_student_grouping.csv';
K = 3;

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(13);
    n = 150;

    CGPA = 5 + 5*rand(n,1);
    Aptitude_Score = randi([40 100], n, 1);
    Coding_Score = randi([35 100], n, 1);
    Attendance = randi([55 100], n, 1);

    data = table(CGPA, Aptitude_Score, Coding_Score, Attendance);
else
    data = readtable(csvFile);
    % Expected columns:
    % CGPA, Aptitude_Score, Coding_Score, Attendance
end

disp(head(data));

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

X = [data.CGPA, data.Aptitude_Score, data.Coding_Score, data.Attendance];
X = normalize(X);

%% =========================
% MODEL BUILDING
% ==========================
[idx, C, sumd] = kmeans(X, K, 'Replicates', 5);

inertia = sum(sumd);

fprintf('Inertia = %.4f\n', inertia);

%% =========================
% SILHOUETTE SCORE
% ==========================
s = silhouette(X, idx);
silhouetteScore = mean(s);
fprintf('Silhouette Score = %.4f\n', silhouetteScore);

%% =========================
% VISUALIZATION
% ==========================
figure;
gscatter(X(:,1), X(:,2), idx);
hold on;
plot(C(:,1), C(:,2), 'kx', 'MarkerSize', 14, 'LineWidth', 3);
xlabel('Normalized CGPA');
ylabel('Normalized Aptitude Score');
title('K-Means Clusters');
grid on;

File: src/data/codes/kmeans.py
"""K-means + silhouette
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_samples, silhouette_score

USE_SYNTHETIC = True
K = 3
RNG = np.random.default_rng(13)

if USE_SYNTHETIC:
    n = 150
    data = pd.DataFrame(
        {
            "CGPA": RNG.uniform(5, 10, size=n),
            "Aptitude_Score": RNG.integers(40, 101, size=n),
            "Coding_Score": RNG.integers(35, 101, size=n),
            "Attendance": RNG.integers(55, 101, size=n),
        }
    )
else:
    data = pd.read_csv("kmeans_student_grouping.csv")

print(data.head())
data = data.dropna().drop_duplicates()
X = data.values.astype(float)
X = StandardScaler().fit_transform(X)

km = KMeans(n_clusters=K, n_init=5, random_state=42)
labels = km.fit_predict(X)
inertia = km.inertia_
print(f"Inertia = {inertia:.4f}")

s = silhouette_samples(X, labels)
sil_avg = silhouette_score(X, labels)
print(f"Silhouette Score = {sil_avg:.4f}")

fig, ax = plt.subplots()
scatter = ax.scatter(X[:, 0], X[:, 1], c=labels, cmap="viridis", alpha=0.7)
ax.scatter(
    km.cluster_centers_[:, 0],
    km.cluster_centers_[:, 1],
    c="k",
    marker="x",
    s=200,
    linewidths=3,
    label="centroids",
)
ax.set_xlabel("Normalized CGPA")
ax.set_ylabel("Normalized Aptitude Score")
ax.set_title("K-Means Clusters")
ax.grid(True)
plt.show()
